rhp/v4/server.go

```diff
@@ -0,0 +1,1078 @@
+			return errorBadRequest("unknown action type %v", action.Type)
+		}
+	}
+
+	resp := rhp4.RPCModifySectorsResponse{
+		Proof: []types.Hash256{rhp4.MetaRoot(roots)}, // TODO implement proof
+	}
+	if err := rhp4.WriteResponse(stream, &resp); err != nil {
+		return fmt.Errorf("failed to write response: %w", err)
+	}
+
+	var renterSigResponse rhp4.RPCModifySectorsSecondResponse
+	if err := rhp4.ReadResponse(stream, &renterSigResponse); err != nil {
+		return errorDecodingError("failed to read renter signature response: %v", err)
+	}
+
+	// revise contract
+	revision, err := rhp4.ReviseForModifySectors(fc, req, resp)
+	if err != nil {
+		return fmt.Errorf("failed to revise contract: %w", err)
+	}
+	sigHash := cs.ContractSigHash(revision)
+
+	// validate the renter signature
+	if !fc.RenterPublicKey.VerifyHash(sigHash, renterSigResponse.RenterSignature) {
+		return rhp4.ErrInvalidSignature
+	}
+	revision.RenterSignature = renterSigResponse.RenterSignature
+	// sign the revision
+	revision.HostSignature = s.hostKey.SignHash(sigHash)
+
+	err = s.contractor.ReviseV2Contract(req.ContractID, revision, roots, Usage{
+		StorageRevenue: cost,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to revise contract: %w", err)
+	}
+	return rhp4.WriteResponse(stream, &rhp4.RPCModifySectorsThirdResponse{
+		HostSignature: revision.HostSignature,
+	})
+}
+
+func (s *Server) handleRPCFundAccounts(stream net.Conn) error {
+	var req rhp4.RPCFundAccountsRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+
+	state, unlock, err := s.lockContractForRevision(req.ContractID)
+	if err != nil {
+		return fmt.Errorf("failed to lock contract: %w", err)
+	}
+	defer unlock()
+
+	var totalDeposits types.Currency
+	for _, deposit := range req.Deposits {
+		totalDeposits = totalDeposits.Add(deposit.Amount)
+	}
+
+	fc := state.Revision
+	if err := rhp4.PayWithContract(&fc, totalDeposits, types.ZeroCurrency); err != nil {
+		return fmt.Errorf("failed to pay with contract: %w", err)
+	}
+
+	sigHash := s.chain.TipState().ContractSigHash(fc)
+	if !fc.RenterPublicKey.VerifyHash(sigHash, req.RenterSignature) {
+		return rhp4.ErrInvalidSignature
+	}
+
+	fc.HostSignature = s.hostKey.SignHash(sigHash)
+
+	balances, err := s.contractor.CreditAccountsWithContract(req.Deposits, req.ContractID, fc)
+	if err != nil {
+		return fmt.Errorf("failed to credit account: %w", err)
+	}
+
+	return rhp4.WriteResponse(stream, &rhp4.RPCFundAccountsResponse{
+		Balances:      balances,
+		HostSignature: fc.HostSignature,
+	})
+}
+
+func (s *Server) handleRPCLatestRevision(stream net.Conn) error {
+	var req rhp4.RPCLatestRevisionRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+
+	state, unlock, err := s.contractor.LockV2Contract(req.ContractID)
+	if err != nil {
+		return fmt.Errorf("failed to lock contract: %w", err)
+	}
+	defer unlock()
```

Comment:
not a big deal, but presumably we can unlock as soon as we observe the `state.Revision`, instead of waiting until after we write the response?

In general we want to avoid holding locks during I/O. Then again, if it only matters when multiple renters are accessing the same contract in parallel, we probably don't need to worry about that.